common/flatpak-dir.c

```diff
@@ -7652,31 +7694,89 @@ flatpak_dir_install (FlatpakDir          *self,
         }
       else
         {
-          /* We're pulling from a remote source, we do the network mirroring pull as a
-             user and hand back the resulting data to the system-helper, that trusts us
-             due to the GPG signatures in the repo */
+          /* If the system-helper can find bindfs on the system, it should return a child repo for us
+             to pull into. The child repo resides inside the system's repo/tmp/flatpak-cache-$ref-XXXXXX
+             directory. This directory is mounted using bindfs which enables us to  pull into it. Once the pull
+             completes, the directory is unmounted before "Deploy" to avoid modifications and do a trusted pull
+             from this child repo to the system repo via hardlinking.
 
-          child_repo = flatpak_dir_create_system_child_repo (self, &child_repo_lock, NULL, error);
+             If this operation is unsupported or failed, the pull is then tried with the fallback codepath. */
+
+          g_autofree gchar *helper_child_repo_path = NULL;
+          g_autoptr(OstreeRepo) child_repo = NULL;
+          g_auto(GLnxLockFile) child_repo_lock = { 0, };
+          g_autoptr(GError) local_error = NULL;
+          gboolean fallback = FALSE;
+
+          if (!flatpak_dir_system_helper_get_child_repo_for_pull (self, ref, opt_commit,
+                                                                  installation ? installation : "",
+                                                                  &helper_child_repo_path, cancellable, &local_error))
+            {
+              if (g_error_matches (local_error, G_DBUS_ERROR, G_DBUS_ERROR_UNKNOWN_METHOD) ||
+                  g_error_matches (local_error, G_DBUS_ERROR, G_DBUS_ERROR_NOT_SUPPORTED))
```

Comment:
@wjt Can you answer this? We had some discussion around it downstream but I am unable to find the comment. AFAIR, you suggested this because a client can be using old versions of system helper..

Reply:
If this is explicitly not supported, we could drop this. The case I was worried about was running the `org.flatpak.Builder` Flatpak with `--install` – couldn't this lead to a case where the unprivileged Flatpak library (from the `org.flatpak.Builder` Flatpak) is newer than the system helper on the host?

Reply:
Hmm, you have a point